### PR TITLE
Allow Routers to inherit state, take 2

### PR DESCRIPTION
**Pull request by jplatte**
## Motivation

Closes https://github.com/tokio-rs/axum/issues/1326.

## Solution

Make `Router<S, _>` no longer a type that *necessarily* holds a state instance of type `S`. Instead, we track (at runtime, to avoid making the generics too complex) whether a given `Router` instance wants to inherit its state.

## Status

I implemented router merging which was much easier than splitting `nest` into `nest` and `nest_service`. The fallback handler (currently stubbed out) will also need some special handling, but I don't think it should be a showstopper.

### CODE REVIEW

**Comment by jplatte** [comment]
@davidpdrsn Have you had a look? I might have some time today or in the coming days to come back to this.

File: axum/src/routing/mod.rs [hunk 1]
```diff
@@ -623,43 +575,52 @@ impl fmt::Debug for Node {
...
+impl<S, B, E> Fallback<S, B, E> {
+    fn map<F, B2, E2>(self, f: F) -> Fallback<S, B2, E2>
     where
-        F: FnOnce(Route<B, E>) -> Route<B2, E2>,
+        S: 'static,
+        B: 'static,
+        E: 'static,
+        F: FnOnce(Route<B, E>) -> Route<B2, E2> + Clone + Send + 'static,
+        B2: 'static,
+        E2: 'static,
```
Comment (davidpdrsn) [inline axum/src/routing/mod.rs:612]: 😂

**Review by davidpdrsn — COMMENTED** [review]
I like the direction of this! I think its a good balance between usability and type safety.

**Comment by davidpdrsn** [comment]
> I implemented router merging which was much easier than splitting nest into nest and nest_service. 

I'm fine with merging this without support for `nest`. Then we can do that in a follow up PR.

File: axum/src/routing/mod.rs [hunk 2]
```diff
@@ -305,29 +322,64 @@ where
...
+                // self has state, attempt to inherit
+                Some(s) => match try_downcast::<Arc<S2>, Arc<S>>(Arc::clone(s)) {
+                    Ok(s) => Some(s),
+                    Err(_) => {
+                        panic!(
+                            "can't merge a `Router` that wants to inherit state of type `{}`\
+                             into a `Router` that holds state of type `{}`",
+                            type_name::<S2>(),
+                            type_name::<S>(),
+                        );
```
[the remark this answers is not in the document]
Comment (jplatte) [reply axum/src/routing/mod.rs:339]: I don't think we talked about it. It's nice to have inference in more cases, but I imagine some people will have use cases where they want different states for different groups of routes, and just not supporting that at all seems unfortunate. Would also be inconsistent with `nest` presumably, or would you expect that to only support one state type as well?

**Comment by jplatte** [comment]
So I basically re-did everything *except* `boxed.rs` with the latest force-push. Things should now be much cleaner.

**Comment by jplatte** [comment]
I also think we can get rid of passing state via `Extension` with the boxed handler approach, but that is a big refactoring on its own that I don't want to mix into this PR.

**Comment by jplatte** [comment]
Rebased to fix the unrelated trybuild failure.

**Comment by jplatte** [comment]
Well, that was a whole lot easier than I expected. I guess some tests would be good, but the implementation makes sense to me as-is.

**Comment by jplatte** [comment]
Added `nest`, it was sooo easy. Hopefully the implementation is correct 😄

**Review by davidpdrsn — COMMENTED** [review]
I think it looks good! I'll write some docs in a follow-up PR.

Wanna update the changelog? Then we can merge this 🚀

**Review by davidpdrsn — APPROVED** [review]

**Pull request merged**